apps/emqx/src/emqx_persistent_session_ds/emqx_persistent_session_ds_state.erl

```diff
@@ -570,17 +904,67 @@ session_iterator_next(Cursor0, N) ->
     ],
     {NextVals, Cursor} = session_iterator_next(mnesia:dirty_next(?session_tab, Cursor0), N - 1),
     {ThisVal ++ NextVals, Cursor}.
+%% END ifdef(STORE_STATE_IN_DS).
+-endif.
 
 %%================================================================================
 %% Internal functions
 %%================================================================================
 
 %% All mnesia reads and writes are passed through this function.
 %% Backward compatiblity issues can be handled here.
+-ifdef(STORE_STATE_IN_DS).
+val_encode(_Domain, Term) ->
+    term_to_binary(Term).
+
+val_decode(_Domain, Bin) ->
+    binary_to_term(Bin).
+
+-spec key_encode(domain(), term()) -> binary().
```

Comment:
Note: key in the DB doesn't necessarily need to be the same as key in pmap. Is it possible to introduce integer handles (created via `get_unique_id`) for objects in PMAP, that will be invisible to the session, but used for (de)serialization?
From my experience, term_to_binary/binary_to_term tends to make unwelcome appearances in the flamegraph all too often.

Reply:
I think we may need to be able to map a key to a topic key deterministically, at least for some cases.

For example, `cold_get_subscription_state` and `cold_get_subscription` directly access specific topics in a session's subscriptions:

https://github.com/thalesmg/emqx/blob/9b7a747c430a2d8c27ad32032e76c3a744664752/apps/emqx/src/emqx_persistent_session_ds/emqx_persistent_session_ds_state.erl#L645-L649

So, if we were to generate unique integers for each subscription or subscription state in the PMAP, I think we would then need to scan all subscriptions / subscription states to find the one matching the argument.

Maybe it's a fine tradeoff, since these functions which are apparently only called from the management APIs?

Reply:
Performance of cold* operations is not very important.

Reply:
Actually, I think we can't generate unique ids for each time we encounter a key.  We need the mapping from pmap key to topic key to be deterministic since we want to overwrite any existing messages with the same key when committing.

Reply:
This mapping can be stored in the pmap structure, perhaps?

Reply:
It can.  But we will still need to be stored alongside the metadata, as the mapping must be reversible.  Otherwise, we won't be able to overwrite existing internal keys after closing and then reopening sessions.